File: R/MonteCarloSimulation.R
rm(list=ls())     # Clean memory
cat("\014")       # Clear Console
graphics.off()    # Close graphs
library(moments)
library(tictoc)

# Load function library
ScriptLocation <- paste0(dirname(rstudioapi::getSourceEditorContext()$path),"/functions/GARCH11Functions.R")
source(ScriptLocation)

#  Simulation parameters for GARCH(1,1) Monte Carlo
theta <- matrix(c(0.1, 0.05, 0.8), nrow=3, ncol=1)
T <- 1E4
Nsim <- 1E3
set.seed(123)

# Initialise matrices
ParaEst <- matrix(NA, nrow=3, ncol=Nsim)
tstat <- matrix(NA, nrow=3, ncol=Nsim)

# Estimate GARCH(1,1) data series
tic()
for (simiter in 1:Nsim)
{
    # Report progress
    if (simiter%%100==0) cat("Iteration", simiter, "out of", Nsim,"\n")
  
    # Generate GARCH(1,1) data set
    data <- GenerateGARCH11(theta, T)
    
    # Estimation
    est <- EstimateGARCH11(data, thetastart = theta)
    ParaEst[, simiter] <- est[[1]]
    tstat[, simiter] <- sqrt(T)*(est[[1]]-theta)/sqrt( diag(est[[3]]) ) # Complex sqrt automatically converted to NaN
}
toc()

SingleMonteCarloRun <- function(theta, T)
{
    # Generate GARCH(1,1) data set
    data <- GenerateGARCH11(theta, T)
    
    # Estimation
    est <- EstimateGARCH11(data, thetastart = theta)
    tstat <- sqrt(T)*(est[[1]]-theta)/sqrt( diag(est[[3]]) ) # Complex sqrt automatically converted to NaN
    return(tstat)
}


tic()
# Implementation using apply (no speed gains to find)
randomSamples <- sapply(c(1:Nsim), function(x) SingleMonteCarloRun(theta, T) )
toc()

# Omit outliers in histogram
SubHistogram <- function(datalist, lowerbound, upperbound)
{
  data2plot <- datalist[lowerbound<datalist & datalist<upperbound]
  hist(data2plot, breaks=20, freq=FALSE, xlim = c(lowerbound, upperbound))
}
  
SubHistogram(tstat[1,], -5, 5)
SubHistogram(tstat[2,], -5, 5)
SubHistogram(tstat[3,], -5, 5)

File: Matlab/MonteCarloSimulation.m
clear variables; clc; close all;
addpath("./functions")

% Simulation parameters for GARCH(1,1) Monte Carlo
theta = [0.1; 0.05; 0.8];
T = 1E4;
Nsim = 1E3;
rng(1)

% Initialise matrices
ParaEst = NaN(3, Nsim);
tstat = NaN(3, Nsim);

tic
% Estimate GARCH(1,1) data series
for simiter = 1:Nsim

    % Report progress
    if (mod(simiter,1E2)) == 0
        fprintf('Iteration %5d out of %5d \n', simiter, Nsim);
    end

    % Generate GARCH(1,1) data set
    data = GenerateGARCH11(theta, T);

    % Estimation
    [est, ~, AsymptCovMatrix] = EstimateGARCH11(data, theta);
    ParaEst(:, simiter) = est;
    tstat(:, simiter) = sqrt(T)*(est-theta)./sqrt( diag(AsymptCovMatrix) );

    % Set complex outcomes to zero (Nb. a flat objective function can cause a negative asymptotic variance)
    for iter = 1:size(ParaEst, 1)
        if ~isreal(tstat(iter, simiter))
            disp("Negative variance encountered")
            tstat(iter, simiter) = NaN;
        end
    end
end
toc

figure(1)
histogram(tstat(2, :), 20, 'Normalization', 'pdf') % Ignores NaN
hold on
xlist = -5:0.05:5;
ylist = normpdf(xlist);
plot(xlist, ylist, 'LineWidth', 2)
hold off
axis([-5 5 0 0.5])

